### ISSUE TITLE
`irsync --ignore-symlinks` against a local directory that contains many symlinks results in `USER_INPUT_PATH_ERR`

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods#7873]
### iRODS Version, OS and Version
4.3.2, almalinux8, iCommands 4.3.2 and ICAT is MySQL

### What did you try to do?
_I see there are some issues with irsync and symlinks (some of them in closed status) but couldn't find a good fitting one to specify this issue hence opened a new one._
 
One of our users tried to synch his local directory that contains quite some symlinks and zero size files to irods. As seen below the command syntax  contains `--ignore-symlinks` as well.

`irsync -v -N 4 --ignore-symlinks -r /staging/leuven/stg_00002/lcb/zkalender/Runs/ i:/gbiomed/home/STAE/users/zkalender/Runs/
`

And the amount of linked files:
```
$ find $LCB/zkalender/Runs -type l|wc -l
59545
```

### Expected behavior
Since no limitation is expressed in the documentation, I expected to see the local dir has been synced to irdos without any error.

### Observed behavior (including steps to reproduce, if applicable)

```
...
20:57:00 JCM__2dc5f1__OmniATAC_K56 0.000 MB --- a match no sync required
20:57:00 JCM__040a60__OmniATAC_RPE 0.006 MB --- a match no sync required
20:57:00 filelist 0.003 MB --- a match no sync required
20:57:00 Mixed_10_MM_lines_sorted. 22209.846 MB --- a match no sync required
20:57:00 remote addresses: 134.58.67.238 ERROR: rsyncUtil: rsync error for /gbiomed/home/STAE/users/zkalender/Runs status = -317000 USER_INPUT_PATH_ERR

```

**Comment by korydraughn** [comment]
It's not clear to me what I'm looking at in the **Observed behavior**. It appears to be client-side messages.

Are there messages in the server log?
Can you provide the minimum set of instructions for reproducing this outside of the user's environment?

**Comment by mstfdkmn** [comment]
Yes that is the client side message (we wrote client messages in a file).

> Are there messages in the server log?

Not sure with the exact timing but I see this error occurred at time as much close to the client error timestamp for the user who executed the `irsync` command.

`[2024-06-25T16:05:01.561Z][icts-p-cloud-rdm-hev-1] {"processname": "irods", "groupname": "irods", "pid": "95", "channel": "stdout", "irods": {"log_category": "legacy", "log_level": "error", "log_message": "iRODS Exception:\n    file: /irods_source/server/core/src/irods_resource_redirect.cpp\n    function: std::string (anonymous namespace)::resolve_hier_for_open_or_write(rsComm_t *, irods::file_object_ptr, const std::string &, const std::string &)\n    line: 245\n    code: -1803000 (HIERARCHY_ERROR)\n    message:\n        no valid resource found for data object\nstack trace:\n--------------\n 0# irods::stacktrace::dump() const in /lib/libirods_common.so.4.3.2\n 1# irods::exception::assemble_full_display_what() const in /lib/libirods_common.so.4.3.2\n 2# irods::exception::what() const in /lib/libirods_common.so.4.3.2\n 3# irods::log(irods::exception const&) in /lib/libirods_common.so.4.3.2\n 4# rsDataObjRsync(RsComm*, DataObjInp*, MsParamArray**) in /lib/libirods_server.so.4.3.2\n 5# irods::api_call_adaptor<DataObjInp*, MsParamArray**>::operator()(irods::plugin_context&, RsComm*, DataObjInp*, MsParamArray**) in /lib/libirods_server.so.4.3.2\n 6# std::__1::__function::__func<irods::api_call_adaptor<DataObjInp*, MsParamArray**>, std::__1::allocator<irods::api_call_adaptor<DataObjInp*, MsParamArray**> >, irods::error (irods::plugin_context&, RsComm*, DataObjInp*, MsParamArray**)>::operator()(irods::plugin_context&, RsComm*&&, DataObjInp*&&, MsParamArray**&&) in /lib/libirods_server.so.4.3.2\n 7# int irods::api_entry::call_handler<DataObjInp*, MsParamArray**>(RsComm*, DataObjInp*, MsParamArray**) in /lib/libirods_server.so.4.3.2\n 8# rsApiHandler(RsComm*, int, BytesBuf*, BytesBuf*) in /lib/libirods_server.so.4.3.2\n 9# readAndProcClientMsg(RsComm*, int) in /lib/libirods_server.so.4.3.2\n10# agentMain(RsComm*) in /lib/libirods_server.so.4.3.2\n11# runIrodsAgentFactory(sockaddr_un) in /lib/libirods_server.so.4.3.2\n12# main::$_5::operator()() const in /usr/sbin/irodsServer\n13# main in /usr/sbin/irodsServer\n14# 0x00007F4A1B116590 in /lib64/libc.so.6\n15# __libc_start_main in /lib64/libc.so.6\n16# _start in /usr/sbin/irodsServer\n\n", "request_api_name": "DATA_OBJ_RSYNC_AN", "request_api_number": 628, "request_api_version": "d", "request_client_user": "u0079808", "request_host": "127.0.0.1", "request_proxy_user": "u0079808", "request_release_version": "rods4.3.2", "server_host": "gbiomed.irods.icts.kuleuven.be", "server_pid": 80549, "server_timestamp": "2024-06-25T16:05:01.561Z", "server_type": "agent", "server_zone": "gbiomed"}}`

To reproduce, this is the command he runs:

`$ irsync -v -N 4 --ignore-symlinks -r /staging/leuven/stg_00002/lcb/zkalender/Runs/ i:/gbiomed/home/STAE/users/zkalender/Runs/ 2>&1 | /staging/leuven/stg_00002/lcb/ghuls/software/moreutils/ts %T | tee /staging/leuven/stg_00002/lcb/ghuls/tmp/zkalender_irsync4.log
`

**Comment by alanking** [comment]
Is this a regression? We just addressed https://github.com/irods/irods/issues/5359 in 4.3.2 which ensures that `--ignore-symlinks` (formerly known as `--link`) ignores symlinks whether they are broken or not. I'm hoping that this was not working before and now is not working. I think our tests cover this case pretty thoroughly, but I wanted to confirm.

Has this user (or you) seen success with a similarly large directory that contains *no* symlinks? What about a very small directory (just a few files) that contains some symlinks?

I'm currently wondering whether the symlinks are the problem or if `irsync` is having some other problem.

**Comment by trel** [comment]
Agreed.   59545 is a lot of symlinks.   Can we show a failure with... one?  or five?  or 100?

**Comment by alanking** [comment]
It has been a long time, but I think we will need a little more to go on in order to reproduce this issue. We have tests for symlinks added in #5359: https://github.com/irods/irods/blob/2b4b38ac6cdd46e4b6fb2aba3d2b0b15cfd7e64e/scripts/irods/test/test_irsync.py#L444-L565 Not sure if there's anything actionable here without more information.

**Comment by alanking** [comment]
Closing this one. Please open a new issue if the issue appears again.
